### PR TITLE
Refactor Data Structures

### PR DESCRIPTION
**Pull request by Dbz**
Ruby
-----------
- [x] binary tree
- [x] doubly linked list
- [x] singly linked list
- [x] linked stack

JavaScript
-----------
- [ ] binary tree
- [ ] hash table
- [ ] queue
- [ ] stack
- [ ] tree
- [x] trie

Cracking the Coding Interview
----------
- [ ] queue with two stacks
- [ ] stack with two queues

### CODE REVIEW

File: Data-Structures/Ruby/binary-tree.rb [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      end
+    end
+  end
+
+  def traverse_dfs &block
+    dfs(@root).map(&:data).compact.each { |data| block.call(data) }
+  end
+
+  def traverse_bfs &block
+    bfs.map(&:data).compact.each { |data| block.call(data) }
```
Comment (isisAnchalee) [inline Data-Structures/Ruby/binary-tree.rb]: This is clever, but it seems computationally inefficient. BFS already *does* traverse each node so why not call the block directly in that?

File: Data-Structures/Ruby/binary-tree.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  private
+
+  def create_root data
+    @root       = Node.new(data)
+    @length     = 1
+    @leaf_nodes = [@root]
+  end
+
+  def dfs node
+    nodes = []
```
Comment (isisAnchalee) [inline Data-Structures/Ruby/binary-tree.rb]: Again, this is clever but is it the most efficient way to write this? Each time you have to concatenate two arrays and then call flat_map... I feel like it's unnecessary.

File: Data-Structures/Ruby/binary-tree.rb [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    dfs(@root).each do |node|
+      if node.data == data
+        @length -= 1
+        node.data = nil
+      end
+    end
+  end
+
+  def traverse_dfs &block
+    dfs(@root).map(&:data).compact.each { |data| block.call(data) }
```
Comment (isisAnchalee) [inline Data-Structures/Ruby/binary-tree.rb]: Same as the other comments. This adds too much unnecessary complexity for the sake of being clever.

**Comment by isisAnchalee** [comment]
Also I think the new structure of the readme is kinda hard to read.

Maybe we should have a section just for Cracking the Coding interview instead of 2 sections in each category.

File: Data-Structures/Ruby/binary-tree.rb (hunk 1 above)
Comment (Dbz) [inline Data-Structures/Ruby/binary-tree.rb]: Good call. Fixed

File: Data-Structures/Ruby/binary-tree.rb (hunk 2 above)
Comment (Dbz) [inline Data-Structures/Ruby/binary-tree.rb]: I don't see a way around concatenation, but I can call dfs earlier to avoid flat_map-ing.

**Comment by Dbz** [comment]
I added a Cracking the Coding Interview section at the end. I was trying to follow how I structured the repo, but maybe I structured the repo incorrectly, and there should be a top level folder for Cracking the Coding Interview.

**Pull request merged**